components/camel-jsonpath/src/main/java/org/apache/camel/jsonpath/JsonPathEngine.java

```diff
@@ -182,14 +182,30 @@ private Object doRead(String path, Exchange exchange) throws IOException, CamelE
             return JsonPath.using(configuration).parse(list).read(path);
         } else {
             // can we find an adapter which can read the message body/header
-            Object answer = readWithAdapter(path, exchange);
+            Object answer = null;
```

Comment:
I would like to see if we can find a solution without catching a runtime exception and then try again with stream mode. 

Instead we should probably check the exchange body and find the right read method to use from the start.

Any processing that causes an exception to be swallowed is a bad smell and also a performance overhead.

Reply:
@davsclaus It makes sense, I'll rework it to decide correctly before an exception is thrown. Thanks for the suggestion.

Reply:
@davclaus I reworked fix to detect `ertx.Buffer` to convert buffer to `InputStream` and handle as an `InputStream.`.

The other option I see is to detect `vertx.Buffer` and use method with InputStream, which uses automatic conversion - which works too. 

But current solution would avoid automatic types convertion and should have a sligfhtly better performance. 

WDYT?

Reply:
The only possible problem could be addition of  dependency `<artifactId>camel-vertx-common</artifactId>`. But I don't see another option how to detect `vertx.Buffer` without it. (so it would be present in both options)

I haven't check karaf feature yet, it has to possible fixed for the current change.

Reply:
We cannot make camel-jsonpath dependent on vert.x. Instead the automatic type converters via Camel is fine to use, to convert the payload to inputstream and use that.

Reply:
ok, thanks for explaining. I'll use automatic conversion.

Reply:
@davsclaus 
There is no way, how to detect that adapter conversion fails in such case (`vertx.Buffer` in this case). Unfortunately default adapter in case of `vertx.Buffer` does not return `null` (which means - can not convert), but return map{"bytes": ... "bytesBuff": ...}, which seems to be a valid conversion with TRACE message.
> JacksonJsonAdapter converted object from: {} to: java.util.Map"

Execution of jsonpath fails afterwards, but without an option to detect why.

The only way of detecting "unknown" content (`vertx.Buffer` in this case) is to try to convert content into `InputStream`. If conversion is successful, we can go with the read method for the input stream.

But there is a small **back-compatibility issue**.

Let say, that there is an adapter, which maps  InputStream into String "InputStream:...text...". 
Current state (before this PR) would return "InpuStream:...text..."
Fix with automatic conversion would change result to "...text...", because automatic conversion would end successfully and would be used before the adapter.

I agree that catching exception is a wrong way, which could cause more problems, then I originally thought of.

So my opinion is:
 1 - try conversion of exchange into inputStream
 2 - if it succeeds, use  inputStream for jsonpath
 3 - if not, use adapter approach as a fallback
This solution brings a slight change in behavior (described above), nevertheless  I think that this should be the correct approach and the difference could be mentioned in upgrade doc. 

If user really needs to use older order (adpter and fallback to inputStream), there are 2 options:
1 - user could remove converter from registry
2 - we can provide a property, which could return component to older behavior
WDYT?

Reply:
Your option is fine. The "problem with the adapter" is that not only introduced more recently with the switch to jackson as json library. The adapter was only something added to make it easier to switch to use jackson instead of that json-smart that it was using as default. But since jackson is now the default, then the adapter is not really to be used.

Just add a note in the 3.10 upgrade guide in the docs about this potential change.

And yes IMHO reading as InputStream first is the best choice - InputStream is used by many components.

Reply:
done